src/spark_rapids_ml/knn.py

```diff
@@ -0,0 +1,362 @@
+    ]:
+        (
+            select_cols,
+            multi_col_names,
+            dimension,
+            feature_type,
+        ) = super()._pre_process_data(dataset)
+
+        select_cols.append(col(alias.row_number))
+
+        return select_cols, multi_col_names, dimension, feature_type
+
+
+class NearestNeighbors(
+    NearestNeighborsClass, _CumlEstimatorNearestNeighbors, _NearestNeighborsCumlParams
+):
+    """
+    Examples
+    --------
+    >>> from spark_rapids_ml.knn import NearestNeighbors
+    >>> data = [([1.0, 1.0],),
+    ...         ([2.0, 2.0],),
+    ...         ([3.0, 3.0],),]
+    >>> topk = 2
+    >>> gpu_knn = NearestNeighbors().setInputCol("features").setK(topk)
+    >>> data_df = spark.createDataFrame(data, ["features"])
+    >>> gpu_knn.fit(data_df)
+    >>> query = [[1.0, 1.0], [3.0, 3.0]]
+    >>> query_df = spark.createDataFrame(query, ["features"])
+    >>> distances_df, indices_df = gpu_knn.kneighbors(query_df)
+    >>> distances_df.show()
+    [0, 1.414]
+    [0, 1.414]
+    >>> indices_df.show()
+    [0, 1]
+    [2, 1]
+    """
+
+    def __init__(self, **kwargs: Any) -> None:
+        super().__init__()
+        self.set_params(**kwargs)
+        self.label_isdata = 0
+        self.label_isquery = 1
+        self.row_number_col = alias.row_number
+        self.set_params(labelCol=alias.label)
+
+    def setK(self, value: int) -> "NearestNeighbors":
+        """
+        Sets the value of `k`.
+        """
+        self.set_params(k=value)
+        return self
+
+    def _create_pyspark_model(self, result: Row) -> "NearestNeighborsModel":
+        return NearestNeighborsModel.from_row(result)
+
+    def fit(self, dataset: DataFrame, params: Optional[Dict[Param[Any], Any]] = None) -> "NearestNeighbors":  # type: ignore
+        self.data_df = dataset.withColumn(alias.label, lit(self.label_isdata))
+        self.data_df = self.df_zip_with_index(self.data_df)
+        return self
+
+    def kneighbors(self, query_df: DataFrame) -> Tuple[DataFrame, DataFrame]:
+        query_default_num_partitions = query_df.rdd.getNumPartitions()
+        query_df = query_df.withColumn(alias.label, lit(self.label_isquery))
+        query_df = self.df_zip_with_index(query_df)
+        union_df = self.data_df.union(query_df)
+
+        pipelinedrdd = self._fit(union_df)
+        pipelinedrdd = pipelinedrdd.repartition(query_default_num_partitions)  # type: ignore
+        res_rdd = pipelinedrdd.flatMap(
+            lambda row: list(zip(row["query_index"], row["indices"], row["distances"]))
+        )
+        res_df = res_rdd.toDF(
+            schema="query_index int, indices array<int>, distances array<float>"
+        )
+        res_df = res_df.sort("query_index")
+        distances_df = res_df.select("distances")
+        indices_df = res_df.select("indices")
+        return (distances_df, indices_df)
+
+    def df_zip_with_index(self, df: DataFrame) -> DataFrame:
```

Comment:
Could you add some comments on this function to describe what's this API doing?

Reply:
Isn't internal.  Probably should be private with `_` prefix.

Reply:
Revised.